tensorpipe/common/cuda_loop.cc

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#include <tensorpipe/common/cuda_loop.h>
+
+#include <tensorpipe/common/cuda.h>
+#include <tensorpipe/common/system.h>
+
+namespace tensorpipe {
+
+CudaLoop::CudaLoop() {
+  std::unique_lock<std::mutex> lock(mutex_);
+  thread_ = std::thread([this, lock{std::move(lock)}]() mutable {
+    setThreadName("TP_CUDA_callback_loop");
+    for (;;) {
+      if (operations_.empty()) {
+        if (joined_ && pendingOperations_ == 0) {
+          break;
+        } else {
+          cv_.wait(lock);
+        }
+      }
+
+      std::deque<Operation> operations;
+      std::swap(operations, operations_);
+
+      lock.unlock();
+      for (auto& op : operations) {
+        op.callback(op.error);
+      }
+      lock.lock();
+
+      pendingOperations_ -= operations.size();
+    }
+  });
+}
+
+CudaLoop::~CudaLoop() {
+  {
+    std::unique_lock<std::mutex> lock(mutex_);
+    joined_ = true;
+    cv_.notify_all();
+  }
+  thread_.join();
+}
+
+void CudaLoop::addCallback(cudaStream_t stream, TCudaCallback callback) {
+  std::unique_lock<std::mutex> lock(mutex_);
+  TP_THROW_ASSERT_IF(joined_);
```

Comment:
I wouldn't be surprised if someone actually tried to add a callback once this is joined (since each channel has its own OnDemandDeferredExecutor, here's what could happen: the context closes, and destroys this loop, and enqueues a close on all channels, but some channels have some other operations already queued up in their on-demand loops, which will run before they "are informed" that they must close. these ops might try to add a callback here, which would fail). Another approach could be to just call the callback inline with some error.